crates/emitter/src/scope_pass.rs

```diff
@@ -0,0 +1,859 @@
+        //
+        // Block : { StatementList }
+        //
+        // Step 1. Let oldEnv be the running execution context's
+        //         LexicalEnvironment.
+        // (implicit)
+
+        // Step 2. Let blockEnv be NewDeclarativeEnvironment(oldEnv).
+        let index = self.scopes.allocate();
+        let context = BlockContext::new(index);
+        self.non_global.insert(block, index);
+
+        // Step 3. Perform
+        //         BlockDeclarationInstantiation(StatementList, blockEnv).
+        // (done in leave_enum_statement_variant_block_statement)
+
+        // Step 4. Set the running execution context's LexicalEnvironment to
+        //         blockEnv.
+        self.context_stack.push_block(context);
+
+        // Step 5. Let blockValue be the result of evaluating StatementList.
+        // (done in runtime)
+    }
+
+    fn leave_enum_statement_variant_block_statement(&mut self, _block: &mut Block<'alloc>) {
+        let context = self.context_stack.pop_block();
+        let enclosing = self.context_stack.current_scope_index();
+
+        // https://tc39.es/ecma262/#sec-block-runtime-semantics-evaluation
+        // Runtime Semantics: Evaluation
+        //
+        // Block : { StatementList }
+        //
+        // Step 3. Perform
+        //         BlockDeclarationInstantiation(StatementList, blockEnv).
+        self.scopes
+            .populate(context.scope_index, context.into_scope_data(enclosing));
+
+        // Step 6. Set the running execution context's LexicalEnvironment to
+        //         oldEnv.
+
+        // Step 7. Return blockValue.
+        // (implicit)
+    }
+
+    fn enter_variable_declaration(&mut self, ast: &mut VariableDeclaration<'alloc>) {
+        match ast.kind {
+            VariableDeclarationKind::Var { .. } => {
+                self.scope_kind_stack.push(ScopeKind::Var);
+            }
+            VariableDeclarationKind::Let { .. } => {
+                self.scope_kind_stack.push(ScopeKind::Let);
+            }
+            VariableDeclarationKind::Const { .. } => {
+                self.scope_kind_stack.push(ScopeKind::Const);
+            }
+        }
+    }
+
+    fn leave_variable_declaration(&mut self, ast: &mut VariableDeclaration<'alloc>) {
+        match ast.kind {
+            VariableDeclarationKind::Var { .. } => {
+                self.scope_kind_stack.pop(ScopeKind::Var);
+            }
+            VariableDeclarationKind::Let { .. } => {
+                self.scope_kind_stack.pop(ScopeKind::Let);
+            }
+            VariableDeclarationKind::Const { .. } => {
+                self.scope_kind_stack.pop(ScopeKind::Const);
+            }
+        }
+    }
+
+    fn visit_binding_identifier(&mut self, ast: &mut BindingIdentifier) {
+        if self.scope_kind_stack.is_empty() {
+            // FIXME
+            // Do nothing for unsupported case.
+            // Emitter will return NotImplemented anyway.
+            return;
+        }
```

Comment:
Would this be enough? What would be the behavior on the following test case:

```js
let a = {b: 1};
```

`visit_binding_identifier` is called for variable declarations, function names and object properties. Using the `scope_kind_stack` would be miss-leading.

You should probably implement the `visit_variable_declarator_binding` function to skip the expression or prevent `visit_binding_identifier` to be miss-lead.

Reply:
`BindingIdentifier` isn't used for object property.
it's used only for object bindings, that is
```
let {b: c} = d;
```

then, about function name, I was about to handle it without stack, given there's at most one, and no need to track the kind with the stack + visit, but just handle it when visiting `Function` node, without visiting `Function.node`.

Reply:
I also saw it being called from `visit_function` and `visit_class_expression`, and many more:
```js
let a = function b () {};
```

https://github.com/mozilla-spidermonkey/jsparagus/blob/ci_generated/crates/ast/src/visit_generated.rs#L5224-L5231

Reply:
yes, I've added comment about those fields that needs special handling.
`ScopePass` should define `visit_*` for those nodes, with function that doesn't visit name fields.